src/analysis/normed_space/hahn_banach/separation.lean

```diff
@@ -0,0 +1,353 @@
+/-
+Copyright (c) 2021 Bhavik Mehta All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Bhavik Mehta, Yaël Dillies
+-/
+import analysis.convex.cone
+import analysis.convex.topology
+import analysis.seminorm
+import tactic.by_contra
+
+/-!
+# Separation Hahn-Banach theorem
+
+In this file we prove the geometric Hahn-Banach theorem. For any two disjoint convex sets, there
+exists a continuous linear functional separating them, geometrically meaning that we can intercalate
+a plane between them.
+
+We provide many variations to stricten the result under more assumptions on the convex sets:
+* `geometric_hahn_banach_open`: One set is open. Weak separation.
+* `geometric_hahn_banach_open_point`, `geometric_hahn_banach_point_open`: One set is open, the
+  other is a singleton. Weak separation.
+* `geometric_hahn_banach_open_open`: Both sets are open. Semistrict separation.
+* `geometric_hahn_banach_compact_closed`, `geometric_hahn_banach_closed_compact`: One set is closed,
+  the other one is compact. Strict separation.
+* `geometric_hahn_banach_point_closed`, `geometric_hahn_banach_closed_point`: One set is closed, the
+  other one is a singleton. Strict separation.
+* `geometric_hahn_banach_point_point`: Both sets are singletons. Strict separation.
+-/
+
+open function set
+open_locale pointwise
+
+variables {𝕜 E : Type*}
+
+section
+open filter
+open_locale topological_space
+
+lemma continuous_at_of_exists_open [normed_ring 𝕜] [normed_group E] [module 𝕜 E]
+  (f : E →ₗ[𝕜] 𝕜) {x : E}
+  (hf : ∀ ε, 0 < ε → ∃ (U : set E), x ∈ U ∧ is_open U ∧ ∀ y ∈ U, ∥f y - f x∥ < ε) :
+  continuous_at f x :=
```

Comment:
I don't think this lemma and the next need to exist, there are a million variants and you can always get the variant you need by invoking the right result about filter bases.  I'll suggest-change a proof showing how in the place where you use this.